Rearchitecture

mxinden opened the pull request:
As discussed in #8 this PR introduces the architecture changes. 

Looking forward to your feedback.

Fixes issue #8 

File tree:
```
├── Alerts
│   ├── Api.elm
│   └── Types.elm
├── Silences
│   ├── Api.elm
│   ├── Decoders.elm
│   ├── Encoders.elm
│   ├── Types.elm
│   └── Views.elm
├── Status
│   ├── Api.elm
│   └── Types.elm
├── Utils
│   ├── Api.elm
│   ├── Date.elm
│   ├── Filter.elm
│   ├── List.elm
│   ├── Types.elm
│   └── Views.elm
├── Views
│   ├── AlertList
│   │   ├── Filter.elm
│   │   ├── Parsing.elm
│   │   ├── Types.elm
│   │   ├── Types.elm~Updated upstream
│   │   ├── Updates.elm
│   │   └── Views.elm
│   ├── NavBar
│   │   └── Views.elm
│   ├── NotFound
│   │   └── Views.elm
│   ├── Shared
│   │   ├── AlertCompact.elm
│   │   ├── AlertListCompact.elm
│   │   ├── SilenceBase.elm
│   │   └── SilencePreview.elm
│   ├── Silence
│   │   ├── Parsing.elm
│   │   ├── Types.elm
│   │   ├── Updates.elm
│   │   └── Views.elm
│   ├── SilenceForm
│   │   ├── Parsing.elm
│   │   ├── Types.elm
│   │   ├── Updates.elm
│   │   └── Views.elm
│   ├── SilenceList
│   │   ├── Parsing.elm
│   │   ├── Types.elm
│   │   ├── Updates.elm
│   │   └── Views.elm
│   └── Status
│       ├── Parsing.elm
│       ├── Types.elm
│       ├── Updates.elm
│       └── Views.elm
├── Main.elm
├── Parsing.elm
├── Types.elm
├── Updates.elm
└── Views.elm

```

w0rm commented on src/Silences/Api.elm:
```diff
@@ -1,36 +1,37 @@
 
-getSilences : Filter -> Cmd Msg
-getSilences filter =
+getSilences : Filter -> (ApiData (List Silence) -> Msg) -> Cmd Msg
```
Api requests do not need to know about the `Msg` type, you can make it lowercase and remove the dependency on the top level `import Types exposing (Msg)`.

stuartnelson3 replied:
Right, if you change it to:

```
getSilences : Filter -> (ApiData (List Silence) -> msg) -> Cmd msg
```

then the caller can supply whatever generic message they want

w0rm commented on src/Silences/Types.elm:
```diff
@@ -90,3 +32,20 @@ nullDuration =
+    }
+
+
+type alias SilenceId =
```
This will not create a new type, so can be substituted with `String`. If you want it to be typechecked, you can do `type SilenceId  = SilenceId String` and then write a decoder that fails for invalid silence ids.

mxinden replied:
Doesn't this make it more descriptive? Instead of handling a `String` in other places we then handle a `SilenceId`? See ELM [Tutorial](https://www.elm-tutorial.org/en/01-foundations/06-type-aliases.html)

w0rm replied:
To me this adds unnecessary noise. The official Elm guide doesn't have this example: https://guide.elm-lang.org/types/type_aliases.html

w0rm commented on src/Status/Api.elm:
```diff
@@ -1,22 +1,22 @@
 
-getStatus : Cmd Msg
-getStatus =
+getStatus : (ApiData StatusResponse -> Msg) -> Cmd Msg
```
Same here `(ApiData StatusResponse -> Msg) -> Cmd Msg` can be written as `(ApiData StatusResponse -> msg) -> Cmd msg` so that `import Types exposing (Msg(MsgForStatus))` may be removed.

w0rm commented on src/Views.elm:
```diff
@@ -2,14 +2,16 @@ module Views exposing (..)
-import Status.Views
-import NavBar.Views exposing (appHeader)
+import Utils.Views exposing (error, loading)
+import Views.SilenceList.Views
```
Not a fan of deep nesting. But if we choose to go for this, we can at least rename the imports `import Views.SilenceList.Views as SilenceList`.

w0rm commented on src/Views.elm:
```diff
@@ -34,21 +36,33 @@ appBody : Model -> Html Msg
+            Views.SilenceList.Views.view model.silences model.silence model.currentTime model.filter
+
+        SilenceFormNewRoute ->
+            Views.SilenceForm.Views.newForm model
```
This would read nicer as `SilenceForm.new model`, right now it says `View` and `Form` twice.

stuartnelson3 replied:
Several of the modules and their submodules introduce some stutter in the naming. I think this is a good suggestion.

mxinden replied:
Yes, this would read a lot nicer. Let me change that.

mxinden replied:
In the long run I would prefer to have separate views for the edit and new which both share the basic input form code but for example have different buttons (e.g. "Create" vs. "Update" buttons).

w0rm commented on src/Views/SilenceForm/Views.elm:
```diff
@@ -0,0 +1,106 @@
+import Html.Attributes exposing (class, href)
+import Html.Events exposing (onClick)
+import Silences.Types exposing (Silence)
+import Types exposing (Msg(MsgForSilenceList, PreviewSilence, MsgForSilenceForm), Model)
```
I don't think we should depend on the top level `Model` type here. `editForm` and `newForm` only require the silence type.

w0rm commented on src/Views/AlertList/Types.elm:
```diff
@@ -0,0 +1,14 @@
+
+
+type Route
+    = Receiver (Maybe String) (Maybe Bool) (Maybe String)
```
Maybe something for later, but this is a nicer type for `Route` because it names the attributes:

```
type alias Route = {
  receiver: Maybe String,
  silenced: Maybe Bool,
  filter: Maybe String
}
```

This parser will work:

```
alertsParser : Parser (Route -> a) a
alertsParser =
  map Route (s "alerts" <?> stringParam "receiver" <?> boolParam "silenced" <?> stringParam "filter")
```

stuartnelson3 replied:
Definitely something to note -- I like how this is self documenting, while in what I originally wrote it's a total mystery what the different `Maybe`s represent.

mxinden replied:
Created issue #13 so this doesn't get lost.

w0rm commented on src/Views/Shared/AlertListCompact.elm:
```diff
@@ -0,0 +1,22 @@
+    let
+        alerts =
+            List.concatMap
+                (\b ->
```
You can use `.alerts` as a shorthand for `(\b -> b.alerts)`

stuartnelson3 replied:
😭 it's so beautiful

mxinden replied:
I think the current way is more descriptive and easier to read. Not a strong opinion. Happy to change it to your suggestion though?!

w0rm commented on src/Views/Shared/AlertListCompact.elm:
```diff
@@ -0,0 +1,22 @@
+        ul
+            [ class "list pa0"
+            ]
+            (List.indexedMap Views.Shared.AlertCompact.view alerts)
```
I would write this function like this:

```
view : AlertGroup -> Html msg
view { blocks } =
  blocks
   |> List.concatMap .alerts
   |> List.indexedMap Views.Shared.AlertCompact.view
   |> ul [ class "list pa0" ]
```

stuartnelson3 replied:
😭 😭 :sob:

mxinden replied:
This is beautiful code!! Forget my comment above. Changed with a5d772752b29a41aaec2fcf798b4e13a8eb3cf1f

w0rm commented on src/Views/Shared/SilenceBase.elm:
```diff
@@ -0,0 +1,50 @@
+                    m.value
+
+                Nothing ->
+                    ""
```
This would be a bit shorter:

```
alertName =
    silence.matchers
      |> List.filter (\m -> m.name == "alertname")
      |> List.head
      |> Maybe.map .value
      |> Maybe.withDefault ""
```

w0rm commented on src/Views/Silence/Updates.elm:
```diff
@@ -0,0 +1,30 @@
+                            Cmd.none
+            in
+                ( { model | silence = s }, cmd )
+
```
This deeply nested case may be split into two top level cases:

```
SilenceFetched (Success sil) ->
    ( { model | silence = Success sil }
    , Task.perform PreviewSilence (Task.succeed sil)
    )

SilenceFetched silence ->
    ( { model | silence = silence }, Cmd.none )
```

mxinden replied:
Done

w0rm commented:
Good work! I've left a few comments, but overall I think we can just merge this.

mxinden commented:
@w0rm Thanks for this big review! I will address your comments later today.

stuartnelson3 commented on src/Alerts/Types.elm:
```diff
@@ -1,29 +1,23 @@
-    | FilterAlerts
+type alias AlertGroup =
+    { blocks : List Block
+    , labels : Utils.Types.Labels
```
Can just be `Labels`

stuartnelson3 commented on src/Parsing.elm:
```diff
@@ -42,19 +43,17 @@ urlParser location =
-topLevelParser : Parser a a
-topLevelParser =
-    s ""
+                Debug.log "inside nothing!" NotFoundRoute
```
Probably can remove the log :)

mxinden replied:
Ou embarrassing. Thanks!

stuartnelson3 commented on src/Updates.elm:
```diff
@@ -0,0 +1,148 @@
+import Navigation
+import Silences.Types exposing (nullSilence)
+import Task
+import Types exposing (Msg(..), Model, Route(NotFoundRoute, SilenceFormEditRoute, SilenceFormNewRoute, SilenceRoute, StatusRoute, SilenceListRoute, AlertsRoute))
```
Not sure what the convention is, but it might be nice to put this on multiple lines

stuartnelson3 commented on src/Updates.elm:
```diff
@@ -0,0 +1,148 @@
+        NavigateToAlerts alertsRoute ->
+            let
+                ( alertsMsg, filter ) =
+                    (Views.AlertList.Updates.urlUpdate alertsRoute)
```
Note to self: This can probably be removed since there's only a single route, the `AlertsGroup` route, being used. When we add `id` to the alerts and have an individual detail page for them, we'll follow the pattern established in this PR and give them their own route and view model.

mxinden replied:
Good idea. Sorry it's too late for me to do it tonight. Can't wrap my head around it right now. I will create an issue so we don't forget. See issue #14.

stuartnelson3 commented on src/Utils/Views.elm:
```diff
@@ -96,7 +89,7 @@ error err =
                 BadPayload err resp ->
                     -- OK status, unexpected payload
-                    "unexpected response from api"
+                    "unexpected response from api" ++ err
```
👍

stuartnelson3 commented on src/Main.elm:
```diff
@@ -37,146 +47,51 @@ init location =
+                NavigateToSilenceList maybeFilter
+
+            SilenceRoute silenceId ->
+                NavigateToSilence silenceId
```
I'll have to check, but we might be able to cut out `NavigateToSilence silenceId` and replace it with `MsgForSilence (InitSilenceView silenceId)`

mxinden replied:
But then we would have to set the new route with the InitSilenceView msg?

stuartnelson3 commented on src/Views/AlertList/Parsing.elm:
```diff
@@ -1,6 +1,6 @@
-module Alerts.Parsing exposing (..)
+module Views.AlertList.Parsing exposing (alertsParser)
```
My initial thought when reading this is "`alertsParser` isn't the responsibility of the view".

I'm thinking it might make more sense for this to live somewhere in the top level app, and then it knows which view to render?

mxinden replied:
So should we combine all Views.*.Parsing into the global Parsing file? Sgtm

stuartnelson3 commented on src/Views/Silence/Updates.elm:
```diff
@@ -0,0 +1,30 @@
+            in
+                ( { model | silence = s }, cmd )
+
+        InitSilenceView silenceId ->
```
This is the same as the `FetchSilence id` branch, so we should be able to just replace all instances of `InitSilenceView silenceId` with that and delete this msg.

mxinden replied:
I thought about introducing this msg for every view. In case there are more steps then just fetching a silence to initlialize the view, then the logic of the initialization steps could be triggered by a single standardized msg in the global update function. What do you think?

stuartnelson3 commented on src/Views/SilenceForm/Updates.elm:
```diff
@@ -0,0 +1,150 @@
+        SilenceCreate silence ->
+            case silence of
+                Success id ->
+                    ( { model | silence = Loading }, Task.perform identity (Task.succeed <| NewUrl ("/#/silence/" ++ id)) )
```
I don't know if `Task.perform identity` is the best way to do this (@w0rm?), but we should definitely move it into a function to shorten all the repetition.

mxinden replied:
Yes for sure. @w0rm Is there a better way of just dispatching a new message in an update function?

w0rm replied:
```
( { model | silence = Loading }, Task.perform NewUrl (Task.succeed ("/#/silence/" ++ id)) )
```

This looks a bit nicer

w0rm replied:
Although Elm folks tend to say that creating a command directly from a message is an anti pattern, I'm not immediately sure what would be a better way to solve this. Maybe something for the future.

stuartnelson3 commented:
One more thing: `send_alerts.sh` and `simple.yml` seem to be some helper files for running alert manager, but don't really pertain to the front end. Could you remove them? (Unless there's a reason you want to keep them)

mxinden commented:
@stuartnelson3 @w0rm Thank you very much for the profound review. Learning a lot here! I have addressed all your comments either by a commit or a comment.

@stuartnelson3 Sorry for the accidentally added files.
